### PR TITLE
feat(zq): Add modular arithmetic primitive with wrapping operations

### LINKED ISSUE #17: Implement Zq - ring of integers modulo q

### PR DESCRIPTION
**Pull request by pycckuu**
## Summary
Added `Zq` struct implementing ℤ/(2³²)ℤ ring arithmetic using native u32 wrapping operations. Provides type-safe modular arithmetic primitives for cryptographic applications.

closes #17 

## Code Changes

**labrador/src/[zq.rs](http://zq.rs/)** (new file)
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Zq {
    value: u32,  // Core storage with automatic wrapping
}

macro_rules! impl_arithmetic {  // DRY operator implementation
    // Generates Add/Sub/Mul traits and Assign variants
}

#[cfg(test)]
mod tests {  // Comprehensive test coverage
    // 8 test cases verifying arithmetic behavior
}
```

## Reason for Changes
Enables safe modular arithmetic required for:
-  Cryptographic primitives (FHE, lattice-based crypto)
-  Constant-time operations using native CPU instructions
-  Prevention of raw integer misuse through type-safe API

## Impact of Changes
-  Zero-cost abstraction: All operations compile to single CPU instructions
-  Memory safety: No heap allocations or branching
-  Cryptographic safety: Wrapping ops prevent timing side channels

## Test Plan
**Implemented tests verify:**
```rust
test_basic_arithmetic()         // Core operations
test_wrapping_arithmetic()      // Overflow/underflow
test_subtraction_edge_cases()   // Boundary conditions
test_assignment_operators()     // In-place operations
test_negative_arithmetic()      // Equivalent negative values
```

## How I tested
-  Validated all test cases pass

## How the reviewer should test
1. Check arithmetic correctness across edge cases:
   - u32::MAX + 1 → 0
   - 1 - u32::MAX → 2
2. Verify trait implementations match spec
3. Confirm test coverage for all operators

## Other useful info
-  Q = 2³² defined as u32::MAX.wrapping_add(1)

### CODE REVIEW

**Comment by NiDimi** [comment]
I think it would be better if we could set the modulus dynamically and not assume its $2^{32}$. It might not make sense now, but I suspect we might need it with our implementation for Falcon with LaBRADOR later. However, @maksimryndin is the Rust god here, so I will let him decide what makes more sense. Nevertheless, we should continue using u32, I don't think its going to be higher.

File: labrador/src/zq.rs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    value: u32,
+}
+
+impl Zq {
+    /// Modulus q = 2^32 (stored as 0 in u32 due to wrapping behavior)
+    pub const Q: u32 = u32::MAX.wrapping_add(1);
+
+    /// Creates a new Zq element from a raw u32 value.
+    /// No explicit modulo needed as u32 automatically wraps
+    pub fn new(value: u32) -> Self {
```
Comment (maksimryndin) [inline labrador/src/zq.rs:17]: In Rust (as well as in C++ constexpt, Zig, Nim etc) we have such a nice feature as compile time evaluation (`const`). Actually, the compiler may evaluate the `const fn` at compile time (if called in a const context) rather than at runtime, reducing the number of instructions.

For example https://godbolt.org/z/rx9s8fEs4 (try to uncomment each line in `main` in turn to see the resulting assembly)

You can also replace `Zq::new(1)` with `Zq::ONE`  in the tests to play around :)

**Review by maksimryndin — APPROVED** [review]
Awesome job Igor! 👍 🔥 

Please consider my comment only for learning purposes. It is not critical at all.

**Comment by maksimryndin** [comment]
> I think it would be better if we could set the modulus dynamically and not assume its 2 32 . It might not make sense now, but I suspect we might need it with our implementation for Falcon with LaBRADOR later. However, @maksimryndin is the Rust god here, so I will let him decide what makes more sense. Nevertheless, we should continue using u32, I don't think its going to be higher.

Nick, thanks for verifying the potential applicability of the solution 👍 

Probably, let's go with the current implementation without a premature generalization. If we get the working prototype, it will be a great result to base a further development. After checking out Falcon aggregation, probably the abstractions requirements may change. 🤷

File: labrador/src/zq.rs (hunk 1 above)
Comment (pycckuu) [reply labrador/src/zq.rs:17]: Thanks! Good to know!

**Comment by mattsuffern** [comment]
Awesome! Thank you for implementing this! I checked the tests, and they seem to cover all operations. I’ll use this struct for polynomial operations then :)

**Review by NiDimi — APPROVED** [review]
LGTM!
